### PR TITLE
Moved to poetry

### PR DESCRIPTION
**Pull request by matacoder**
Resolves: #64
Resolves: #60 

How to work with poetry:
0. Get [poetry](https://python-poetry.org/docs/#installation)
1. run `poetry install` to install everything you need to develop
2. run `poetry run blackbox` to start the program
3. use `poe lint` for linting, `poe test` for testing and `poe isort` for import sorting.

`Dockerfile` updated and blackbox run flawless within containers.

Additional feature: semver for dockerhub directly from Poetry TOML file. One and only place to keep version number.
Need to check double echo output in workflow!

+++ Added workflow to publish to PyPI using Poetry +++
Need to double check

### CODE REVIEW

**Comment by matacoder** [comment]
And Github linter must be updated to run with poetry, `pipenv` files are no longer there

File: .github/workflows/build_and_push.yaml [hunk 1]
```diff
@@ -45,4 +47,5 @@ jobs:
           cache-to: type=inline
           tags: |
             lemonsaurus/blackbox:latest
-            lemonsaurus/blackbox:${{ steps.sha_tag.outputs.tag }}
+            lemonsaurus/blackbox:${{ steps.tags_generator.outputs.tag }}
+            lemonsaurus/blackbox:${{ steps.tags_generator.outputs.semver }}
```
Comment (lemonsaurus) [inline .github/workflows/build_and_push.yaml:51]: This looks like the right approach, but this workflow executes on push to main, which makes it inaccurate. We can't push a semver-tagged version to DockerHub every time there's a master commit - We have to push them to DockerHub when we actually _release_ a new version, via the GitHub Release mechanism.

File: .github/workflows/publish.yaml [hunk 2]
```diff
@@ -2,10 +2,10 @@ name: Release to PyPI
 
 on:
   release:
-    types: [published]
+    types: [ published ]
 
 jobs:
-  build:
+  publish-pipy:
```
Comment (lemonsaurus) [inline .github/workflows/publish.yaml:8]:
```suggestion
  publish-to-pipy:
```
Maybe like this. Nice find, though.

File: .github/workflows/tests.yaml [hunk 3]
```diff
@@ -5,31 +5,38 @@ name: Lint & Test
...
 jobs:
-  build:
-    runs-on: ubuntu-latest
+  test-lint:
     strategy:
+      fail-fast: false
       matrix:
-        python-version: [3.9]
+        python-version: [ 3.9.4 ]
+        poetry-version: [ 1.1.5 ]
```
Comment (lemonsaurus) [inline .github/workflows/tests.yaml:18]: This seems strange. Why do we need a matrix strategy to test a single version?

File: tox.ini [hunk 4]
```diff
@@ -10,7 +10,7 @@ ignore = P102,B311,W503,E226,S311, E221
          D301,D302,
          # Docstring Content
          D401,D402,D404,D405,D406,D407,D408,D409,D410,D411,D412,D413,D414,D416,D417
-exclude = __pycache__, venv, .venv, .git, setup.py, __init__.py
+exclude = __pycache__, venv, .venv, .git, setup.py, __init__.py, act
```
Comment (lemonsaurus) [inline tox.ini:13]: what's `act` ?

File: docker/Dockerfile [hunk 5]
```diff
@@ -1,27 +1,42 @@
...
 # and redis-tools, used for redis backups.
 RUN apt-get install -y mongo-tools redis-tools
 
 # Install the Postgres 12 client, needed for pg_dumpall
+# And MariaDB client for mysqldump
 RUN sh -c 'echo "deb http://apt.postgresql.org/pub/repos/apt $(lsb_release -cs)-pgdg main" > /etc/apt/sources.list.d/pgdg.list'
 RUN wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
 RUN apt-get update -y && \
-    apt-get install -y postgresql-client-12
+    apt-get install -y postgresql-client-12 mariadb-client
```
Comment (lemonsaurus) [inline docker/Dockerfile:29]: slightly weird PR to add this in, but it's okay. Won't do any damage, and we want it anyway.

File: docker/Dockerfile [hunk 6]
```diff
@@ -1,27 +1,42 @@
...
+
 # Install essential tools
 RUN apt-get -y update && apt-get install -y \
     wget \
     gnupg \
     lsb-release
 
+# Install and setup poetry
+RUN pip install -U pip \
+    && apt install -y curl netcat \
```
Comment (lemonsaurus) [inline docker/Dockerfile:16]: we really need `netcat` ? what are we using it for?

File: .github/workflows/tests.yaml [hunk 7]
```diff
@@ -5,31 +5,38 @@ name: Lint & Test
...
+          poetry-version: ${{ matrix.poetry-version }}
+
+      - name: Install poetry dependencies
+        run: poetry install --no-interaction --no-root
+
+      - name: Install poethepoet to use poe
+        run: pip install poethepoet
+
+      - name: Linting
+        run: poe lint
```
Comment (lemonsaurus) [inline .github/workflows/tests.yaml:39]: these are nice. I didn't know about poethepoet until now, it looks very nice.

**Review by lemonsaurus — CHANGES_REQUESTED** [review]
Need to push to DH only when it makes sense, but otherwise this is pretty good.

I also should probably test this before I merge it.

File: tox.ini (hunk 4 above)
Comment (matacoder) [reply tox.ini:13]: Act is that offline CI testing tool you advised me to use. It makes a lot of badly linted files.

File: docker/Dockerfile (hunk 5 above)
Comment (matacoder) [reply docker/Dockerfile:29]: Poetry would be the last PR in a row so it should contain all dependencies. I was thinking in advance.

File: .github/workflows/tests.yaml (hunk 3 above)
Comment (matacoder) [reply .github/workflows/tests.yaml:18]: That is approach from Action example https://github.com/abatilo/actions-poetry
I've just used it as is. If it is terrible I can always fix it but to tell the truth I like to see all versions in one section that provides a fast glimpse at the project

File: docker/Dockerfile (hunk 6 above)
Comment (matacoder) [reply docker/Dockerfile:16]: Valid question. It seems to be that we are not using it.

File: .github/workflows/build_and_push.yaml [hunk 8]
```diff
@@ -45,4 +45,4 @@ jobs:
           cache-to: type=inline
           tags: |
             lemonsaurus/blackbox:latest
```
Comment (lemonsaurus) [inline .github/workflows/build_and_push.yaml:47]: I think the `latest` tag should only refer to the latest _release version_, not the latest commit. That is the normal convention.

So, let's tag with `latest` when we do the semver release, and then here we can tag with `main` instead.

**Review by lemonsaurus — CHANGES_REQUESTED** [review]
One small remaining change, and then I think we can merge this.

File: .github/workflows/build_and_push.yaml (hunk 8 above)
Comment (matacoder) [reply .github/workflows/build_and_push.yaml]: Reasonable. Fixed.

**Review by lemonsaurus — APPROVED** [review]
Okay, I think we're where we need to be with this, then. Let's merge it!

**Pull request merged**